crates/factors-executor/src/lib.rs

```diff
@@ -282,6 +328,42 @@ pub struct InstanceState<T, U> {
     core: spin_core::State,
     factors: T,
     executor: U,
+    /// The last time guest code started running in this instance.
+    pub cpu_time_last_entry: Option<Instant>,
+    /// The total CPU time elapsed actively running guest code in this instance.
+    pub cpu_time_elapsed: Duration,
+    /// The component ID.
+    pub component_id: String,
+    /// The memory (in bytes) consumed on initialization.
+    pub mem_usg_init: u64,
+}
+
+impl<T, U> Drop for InstanceState<T, U> {
+    fn drop(&mut self) {
+        // Record the component execution time.
+        #[cfg(feature = "call-hook")]
+        spin_telemetry::metrics::histogram!(
+            spin.component_cpu_time = self.cpu_time_elapsed.as_micros() as f64,
+            component_id = self.component_id,
+            // According to the OpenTelemetry spec, instruments measuring durations should use "s" as the unit.
+            // See https://opentelemetry.io/docs/specs/semconv/general/metrics/#units
+            unit = "s"
+        );
+
+        // Record the component memory consumed on initialization.
+        spin_telemetry::metrics::histogram!(
+            spin.component_init_memory_used = self.mem_usg_init,
+            component_id = self.component_id,
+            unit = "bytes"
```

Comment:
```suggestion
            unit = "By"
```

...apparently.

OpenTelemetry recommends using [UCUM units](https://ucum.org/ucum) (the "c/s" column). This is a nicer reference and shows one place where it actually matters: https://github.com/open-telemetry/opentelemetry-rust/blob/07bb613d0601589e03cb220f139e967eb317923d/opentelemetry-prometheus/src/utils.rs#L55